kernel/libs/aster-bigtcp/src/socket/bound.rs

```diff
@@ -120,51 +368,177 @@ impl<E> AnyBoundSocketInner<E> {
         }
     }
 
-    pub(crate) fn is_closed(&self) -> bool {
-        match self.socket_family {
-            SocketFamily::Tcp => self.raw_with(|socket: &mut RawTcpSocket| {
-                socket.state() == smoltcp::socket::tcp::State::Closed
-            }),
-            SocketFamily::Udp => true,
+    /// Returns the next polling time.
+    ///
+    /// Note: a zero means polling should be done now and a `u64::MAX` means no polling is required
+    /// before new network or user events.
+    #[inline]
+    pub(crate) fn next_poll_at_ms(&self) -> u64 {
+        self.next_poll_at_ms.load(Ordering::Relaxed)
+    }
+
+    /// Updates the next polling time according to `poll_at`.
+    ///
+    /// The update is typically needed after new network or user events have been handled, so this
+    /// method also marks that there may be new events, so that the event observer provided by
+    /// [`BoundSocket::set_observer`] can be notified later.
+    fn update_next_poll(&self, poll_at: PollAt) {
+        self.has_new_events.store(true, Ordering::Relaxed);
+
+        match poll_at {
+            PollAt::Now => self.next_poll_at_ms.store(0, Ordering::Relaxed),
+            PollAt::Time(instant) => self
+                .next_poll_at_ms
+                .store(instant.total_millis() as u64, Ordering::Relaxed),
+            PollAt::Ingress => self.next_poll_at_ms.store(u64::MAX, Ordering::Relaxed),
         }
     }
+}
+
+impl<T, E> BoundSocketInner<T, E> {
+    #[inline]
+    pub(crate) fn port(&self) -> u16 {
+        self.port
+    }
+}
 
-    /// Starts closing the socket and returns whether the socket is closed.
+impl<E> BoundTcpSocketInner<E> {
+    /// Returns whether the TCP socket is dead.
     ///
-    /// For sockets that can be closed immediately, such as UDP sockets and TCP listening sockets,
-    /// this method will always return `true`.
+    /// A TCP socket is considered dead if and only if the following two conditions are met:
+    /// 1. The TCP connection is closed, i.e., the TCP state is in the closed state.
+    /// 2. The socket file is closed, i.e., the socket state is in the closed state.
+    #[inline]
+    pub(crate) fn is_dead(&self) -> bool {
+        self.socket.is_dead()
+    }
```

Comment:
The `is_dead` and `is_dropped` makes me confusing when reading the code. Can you merge them into one state. For example

```rust
pub enum CloseState {
    /// Not closed yet.
    Alive,
    /// The socket file is closed, while the socket is not. 
    /// (This can only happen for TCP sockets)
    Closing, 
    /// The socket file and socket status is both closed
    Dead,
}
```

Reply:
This PR splits the large `SocketSet` lock into specific locks on each `Socket`. However, one of the main goals of this PR is to avoid unnecessary socket locking during polling.

So `is_dead` and `is_dropped` are deliberately designed to ensure that `is_dead` can be loaded without holding the socket lock. Otherwise we would have to pay the locking overhead during `remove_dead_tcp_sockets`, which I don't think is reasonable.

Reply:
Makes sense to me.